pkg/stub/retry.go

```diff
@@ -0,0 +1,147 @@
+	finalTag, existing, multiple, err := followTagReferenceV1(stream, tag)
+	if err != nil {
+		return nil, err
+	}
+
+	if existing == nil || existing.From == nil {
+		return nil, nil
+	}
+
+	// disallow re-importing anything other than DockerImage
+	if existing.From != nil && existing.From.Kind != "DockerImage" {
+		return nil, fmt.Errorf("tag %q points to existing %s %q, it cannot be re-imported", tag, existing.From.Kind, existing.From.Name)
+	}
+
+	// set the target item to import
+	if multiple {
+		tag = finalTag
+	}
+
+	// reset the generation
+	zero := int64(0)
+	existing.Generation = &zero
+
+	// and create accompanying ImageStreamImport
+	return newImageStreamImportTags(stream, map[string]string{tag: existing.From.Name}), nil
+}
+
+func followTagReferenceV1(stream *imagev1.ImageStream, tag string) (finalTag string, ref *imagev1.TagReference, multiple bool, err error) {
+	seen := sets.NewString()
+	for {
+		if seen.Has(tag) {
+			// circular reference; should not exist with samples but we sanity check
+			// to avoid infinite loop
+			return tag, nil, multiple, fmt.Errorf("circular reference stream %s tag %s", stream.Name, tag)
+		}
+		seen.Insert(tag)
+
+		var tagRef imagev1.TagReference
+		for _, t := range stream.Spec.Tags {
+			if t.Name == tag {
+				tagRef = t
+				break
+			}
+		}
+
+		if tagRef.From == nil || tagRef.From.Kind != "ImageStreamTag" {
+			// terminating tag
+			return tag, &tagRef, multiple, nil
+		}
+
+		// The reference needs to be followed with two format patterns:
+		// a) sameis:sometag and b) sometag
+		if strings.Contains(tagRef.From.Name, ":") {
+			tagref := splitImageStreamTag(tagRef.From.Name)
+			// sameis:sometag - follow the reference as sometag
+			tag = tagref
+		} else {
+			// sometag - follow the reference
+			tag = tagRef.From.Name
+		}
+		multiple = true
+	}
+}
+
+func newImageStreamImportTags(stream *imagev1.ImageStream, tags map[string]string) *imagev1.ImageStreamImport {
+	isi := newImageStreamImport(stream)
+	for tag, from := range tags {
+		var insecure, scheduled bool
+		oldTagFound := false
+		var oldTag imagev1.TagReference
+		for _, t := range stream.Spec.Tags {
+			if t.Name == tag {
+				oldTag = t
+				oldTagFound = true
+				break
+			}
+		}
+
+		if oldTagFound {
+			insecure = oldTag.ImportPolicy.Insecure
+			scheduled = oldTag.ImportPolicy.Scheduled
+		}
+		isi.Spec.Images = append(isi.Spec.Images, imagev1.ImageImportSpec{
+			From: corev1.ObjectReference{
+				Kind: "DockerImage",
+				Name: from,
+			},
+			To: &corev1.LocalObjectReference{Name: tag},
+			ImportPolicy: imagev1.TagImportPolicy{
+				Insecure:  insecure,
+				Scheduled: scheduled,
+			},
+			ReferencePolicy: getReferencePolicy(),
+		})
+	}
+	return isi
+}
+
+func newImageStreamImport(stream *imagev1.ImageStream) *imagev1.ImageStreamImport {
+	isi := &imagev1.ImageStreamImport{
+		ObjectMeta: kapis.ObjectMeta{
+			Name:            stream.Name,
+			Namespace:       "openshift",
```

Comment:
do we not have a constant for this("samples installation namespace") somewhere?

Reply:
alternatively can't you get the namespace from stream.Namespace?

Reply:
switched to `stream.Namespace`